alluxio/retry.py

```diff
@@ -0,0 +1,151 @@
+
+        # This is basically deconstructing the HttpError when pickled.
+        return HttpError, (self._error_response,)
+
+
+RETRIABLE_EXCEPTIONS = (
+    requests.exceptions.ChunkedEncodingError,
+    requests.exceptions.ConnectionError,
+    requests.exceptions.ReadTimeout,
+    requests.exceptions.Timeout,
+    requests.exceptions.ProxyError,
+    requests.exceptions.SSLError,
+    requests.exceptions.ContentDecodingError,
+    aiohttp.client_exceptions.ClientError,
+)
+
+
+def is_retriable(exception):
+    """Returns True if this exception is retriable."""
+    errs = list(range(500, 505)) + [
+        # Request Timeout
+        408,
+        # Too Many Requests
+        429,
+    ]
+    errs += [str(e) for e in errs]
+    if isinstance(exception, HttpError):
+        return exception.code in errs
+
+    return isinstance(exception, RETRIABLE_EXCEPTIONS)
+
+
+def validate_response(status, content, path, args=None):
+    """
+    Check the requests object r, raise error if it's not ok.
+
+    Parameters
+    ----------
+    r: requests response object
+    path: associated URL path, for error messages
+    """
+    if status >= 400 and status != 499:
+        # 499 is special "upload was cancelled" status
+        if args:
+            from .core import quote
+
+            path = path.format(*[quote(p) for p in args])
+        if status == 404:
+            raise FileNotFoundError(path)
+
+        error = None
+        if hasattr(content, "decode"):
+            content = content.decode()
+        try:
+            error = json.loads(content)["error"]
+            msg = error["message"]
+        except json.decoder.JSONDecodeError:
+            msg = content
+
+        if status == 403:
+            raise OSError(f"Forbidden: {path}\n{msg}")
+        elif status == 502:
+            raise requests.exceptions.ProxyError()
+        elif "invalid" in str(msg):
+            raise ValueError(f"Bad Request: {path}\n{msg}")
+        elif error:
+            raise HttpError(error)
+        elif status:
+            raise HttpError({"code": status, "message": msg})  # text-like
+        else:
+            raise RuntimeError(msg)
+
+
+@decorator
+async def retry_request(func, retries=1, *args, **kwargs):
+    for retry in range(retries):
+        try:
+            if retry > 0:
+                await asyncio.sleep(
+                    min(random.random() + 2 ** (retry - 1), 32)
+                )
+            return await func(*args, **kwargs)
+        except (
+            HttpError,
+            requests.exceptions.RequestException,
+            aiohttp.client_exceptions.ClientError,
+        ) as e:
+            if (
+                isinstance(e, HttpError)
+                and e.code == 400
+                and "requester pays" in e.message
+            ):
+                msg = (
+                    "Bucket is requester pays. "
+                    "Set `requester_pays=True` when creating the GCSFileSystem."
+                )
```

Comment:
Do we need this?

Reply:
Probably need to design retry in the whole workload perspective
- fsspec, if alluxio-py failed, will fallback to UFS
- alluxio-py, ? should we retry only network connection issue here?
- alluxio RESTful API, does it retry and should it retry retryable internal errors?
- Alluxio internal components

Reply:
I feel like with retryable network issue we can retry otherwise we fallback to UFS